[KOGITO-7481] - [KWS-Guides] Document the Enhanced Service Discovery (#177)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/enhanced-service-discovery.adoc

> = Enhanced Service Discovery

Reviewer: ```suggestion
= Kubernetes Service Discovery for Workflow Application Configuration
```

:compat-mode!:
// Metadata:
:description: Explain what is and how the service discovery works
:keywords: kogito, workflow, quarkus, serverless, service-discovery, enhanced-service-discovery
// links
:quarkus_issue_url: https://github.com/quarkusio/quarkus/issues/27457

> The Enhanced Service Discovery allows you to have a static URI that will define the Kubernetes resource that you use

Reviewer 2: Which property are we talking about ? From Service orchestration ?

> The Enhanced Service Discovery allows you to have a static URI that will define the Kubernetes resource that you use

Reviewer: ```suggestion
The Kubernetes Service Discovery allows you to have a static URI that will define the Kubernetes resource that you use
```

> to perform http requests, the following sketch demonstrate how the SWUrl is composed:

Reviewer 2: what is SWUrl ? Can we call it `Kubernetes Serverless Workflow URI` ?

Reviewer: Please call it just `URI`.

[source,shell]
----
kubernetes:<group>/<version>/<kind>/<namespace>/<name>?<attributeName>=<attributeValue>
\________/ \_____/ \_______/ \____/ \_________/ \____/ \______________________________/
  scheme    group   version   kind   namespace   name   additional resource attributes
            \____________________/                       \__________________________/

>                      GVK                                  Supported value: port_name

Reviewer 2: ```suggestion
                Kubernetes GVK                                  Supported value: port_name
```

----

> On the scheme, the following values are supported:

Reviewer 2: ```suggestion
The following scheme values are supported:
```

* **kubernetes**
* **openshift**
* **knative**

> As GVK, for now, the following Kubernetes Resources are supported:

Reviewer 2: ```suggestion
For the GVK, the following Kubernetes Resources are currently supported:
```

* **v1/service**
* **serving.knative.dev/v1/service**
* **v1/pod**
* **apps/v1/deployment**
* **apps.openshift.io/v1/deploymentconfig**
* **apps/v1/statefulset**
* **route.openshift.io/v1/route**
* **networking.k8s.io/v1/ingress**

> [TIP]
> ====
> Note that the _GVK_(Group, Version and Kind) and the _Resource Name_ are mandatory and can't be empty, otherwhise, the engine will throw a validation issue and the discovery will not be performed. The **namespace** is not mandatory, however, if empty, the current namespace or context will be used.
> ====

Reviewer 2: ```suggestion
[IMPORTANT]
====
The **GVK** (Group, Version and Kind) and the **Name** values are **mandatory** and can't be empty, otherwise, the engine will throw a validation issue and the discovery will not be performed.  +
The **namespace** is **optional**. In case of an empty value, the current namespace or context will be used.
====
```

== Configuration

> There are two **build time** system properties, which are:

Reviewer 2: ```suggestion
You can configure the Service Discovery with two **build time** properties, which are:
```

> * `quarkus.kogito.sw.discovery.enabled`: Disable the service discovery. Defaults to _true_.

Reviewer 2: ```suggestion
* `quarkus.kogito.sw.discovery.enabled`: Enable the service discovery. Defaults to _true_.
```

> * `quarkus.kogito.kubernetes-client.okhttp.static.logger`: Weather or not to print the `http`communication between the client and the Kubernetes API. Defaults to _false_.
> **  Note that, this system property can change in the future since this feature will be ported to the _quarkus-kubernetes-client_ extension. For more info please refer to this link:{quarkus_issue_url}[link].

Reviewer 2: ```suggestion
* `quarkus.kogito.kubernetes-client.okhttp.static.logger`: Whether or not to print the `http` communication between the client and the Kubernetes API. Defaults to _false_. +
_This property may change in the future since this feature will be ported to the _quarkus-kubernetes-client_ extension. For more info please refer to this link:{quarkus_issue_url}[link]._
```

> == Startup time: discovery enabled x disabled

Reviewer 2: I would put this part at the end of the guide and rename it to something like: `Drawbacks to take into account` and be a bit more verbose on why this difference at startup time.

[tabs]
====
Service Discovery Enabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
----
Service Discovery Disabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.507s. Listening on: http://0.0.0.0:8080
----
====

[IMPORTANT]
====
When using this feature, you need to balance if the delayed startup time is something
that your application can afford.
====

== How does it work?

> The service discovery is performed at the Quarkus _STATIC_INIT_ time, which means, before the application gets started. It scans the system properties and searches for the _SWUrl_ pattern, if at least one is found, the engine will parse the SWUrl and then query the Kubernetes API for the given resource and override the given system property. Let's see a small example of it:

Reviewer 3: system property terminology might be confusing. 
In Java, a system property is a property passed to the JVM, usually through command line, using -D prefix. 
What you are actually doing in the code is scanning Quarkus configuration, of which Java system properties are just a subset. 
Hence I think the term quarkus configuration (and a link to Quarkus configuration guide) will be better in that context

> The service discovery is performed at the Quarkus _STATIC_INIT_ time, which means, before the application gets started. It scans the system properties and searches for the _SWUrl_ pattern, if at least one is found, the engine will parse the SWUrl and then query the Kubernetes API for the given resource and override the given system property. Let's see a small example of it:

Reviewer 2: `_SWUrl_` => Kubernetes Serverless Workflow URI

Suppose we have an application that will consume a resource running on Kubernetes, this resource is a Knative Service that exposes a function, that can be discovered with the following SWUrl:

[source,shell]
----
org.kie.kogito.sw.knative.service=knative:v1/Service/serverless-workflow-greeting-quarkus/greeting-quarkus-cli
----

[NOTE]
====
Note that, the system property does not really matter, the discovery is all about the property value.
====

> When the application is started, you should be able to see in the logs the Service Discovery in action:

Reviewer 2: ```suggestion
Once the application has started, you should be able to see in the logs the Service Discovery into action:
```

2022-08-23 12:21:06,375 INFO  [org.kie.kog.sw.AppLifecycle] (main) The application is starting
2022-08-23 12:21:06,382 INFO  [org.kie.kog.add.qua.mes.com.QuarkusKogitoExtensionInitializer] (main) Registered Kogito CloudEvent extension
2022-08-23 12:21:06,460 INFO  [io.quarkus] (main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Profile prod activated.
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Installed features: [cache, cdi, hibernate-validator, jackson-jq, kogito-addon-kubernetes-extension, kogito-addon-messaging-extension, kogito-processes, kogito-serverless-workflow, kubernetes, kubernetes-client, openshift-client, qute, reactive-routes, rest-client, rest-client-jackson, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-openapi, smallrye-reactive-messaging, smallrye-reactive-messaging-http, vertx]
----

[NOTE]
====

> Note that the SWUrl was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io`, and, when the sytem property is used at runtime, the new value will be used.

Reviewer 2: `SWUrl` => `Kubernetes Serverless Workflow URI`

> Note that the SWUrl was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io`, and, when the sytem property is used at runtime, the new value will be used.

Reviewer 2: ```suggestion
Note that the SWUrl was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` when the application started, and will be used at runtime any time it needs it.
```

= Kubernetes Service Discovery for Workflow Application Configuration
:compat-mode!:
// Metadata:
:description: Explain what is and how the service discovery works
:keywords: kogito, workflow, quarkus, serverless, service-discovery, enhanced-service-discovery
// links
:quarkus_issue_url: https://github.com/quarkusio/quarkus/issues/27457

The Kubernetes Service Discovery allows you to have a static URI that will define the Kubernetes resource that you use

> to perform http requests, the following sketch demonstrate how the SWUrl is composed:

Reviewer: ```suggestion
to perform HTTP requests, the following sketch demonstrates how this URI is composed:
```

[source,shell]
----
kubernetes:<group>/<version>/<kind>/<namespace>/<name>?<attributeName>=<attributeValue>
\________/ \_____/ \_______/ \____/ \_________/ \____/ \______________________________/
  scheme    group   version   kind   namespace   name   additional resource attributes
            \____________________/                       \__________________________/
                Kubernetes GVK                                  Supported value: port_name
----

The following scheme values are supported:

* **kubernetes**
* **openshift**
* **knative**

For the GVK, the following Kubernetes Resources are currently supported:

* **v1/service**
* **serving.knative.dev/v1/service**
* **v1/pod**
* **apps/v1/deployment**
* **apps.openshift.io/v1/deploymentconfig**
* **apps/v1/statefulset**
* **route.openshift.io/v1/route**
* **networking.k8s.io/v1/ingress**

[IMPORTANT]
====

> The **GVK** (Group, Version and Kind) and the **Name** values are **mandatory** and can't be empty, otherwise, the engine will throw a validation issue and the discovery will not be performed.  +

Reviewer: I think we can have "special" cases where the engine should assume the resource type. For Kubernetes/OpenShift, `deployment`, and for Knative, `ksvc`. 

Can you open a follow-up JIRA?

Author: That's might be a good idea :)
https://issues.redhat.com/browse/KOGITO-7863

====

== Configuration

You can configure the Service Discovery with two **build time** properties, which are:

* `quarkus.kogito.sw.discovery.enabled`: Disable the service discovery. Defaults to _true_.
* `quarkus.kogito.kubernetes-client.okhttp.static.logger`: Weather or not to print the `http`communication between the client and the Kubernetes API. Defaults to _false_.
**  Note that, this system property can change in the future since this feature will be ported to the _quarkus-kubernetes-client_ extension. For more info please refer to this link:{quarkus_issue_url}[link].

== Startup time: discovery enabled x disabled

[tabs]
====
Service Discovery Enabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
----
Service Discovery Disabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.507s. Listening on: http://0.0.0.0:8080
----
====

[IMPORTANT]
====
When using this feature, you need to balance if the delayed startup time is something
that your application can afford.
====

== How does it work?

> The service discovery is performed at the Quarkus _STATIC_INIT_ time, which means, before the application gets started. It scans the system properties and searches for the _SWUrl_ pattern, if at least one is found, the engine will parse the SWUrl and then query the Kubernetes API for the given resource and override the given system property. Let's see a small example of it:

Reviewer: ```suggestion
During application startup, the service discovery is performed at the Quarkus _STATIC_INIT_ time. First, it scans the Quarkus configuration and searches for the Kubernetes URI pattern. Then if at least one is found, the engine will parse the URI, query the Kubernetes API for the given resource, and override the given system property. Let's see a small example of it:
```
Please link the `Kubernetes URI` to the section you're explaining about the URI structure.

Suppose we have an application that will consume a resource running on Kubernetes, this resource is a Knative Service that exposes a function, that can be discovered with the following SWUrl:

[source,shell]
----
org.kie.kogito.sw.knative.service=knative:v1/Service/serverless-workflow-greeting-quarkus/greeting-quarkus-cli
----

[NOTE]
====

> Note that, the system property does not really matter, the discovery is all about the property value.

Reviewer: ```suggestion
Note that, the application property does not matter. The discovery is all about the property value.
```

2022-08-23 12:21:06,460 INFO  [io.quarkus] (main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Profile prod activated.
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Installed features: [cache, cdi, hibernate-validator, jackson-jq, kogito-addon-kubernetes-extension, kogito-addon-messaging-extension, kogito-processes, kogito-serverless-workflow, kubernetes, kubernetes-client, openshift-client, qute, reactive-routes, rest-client, rest-client-jackson, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-openapi, smallrye-reactive-messaging, smallrye-reactive-messaging-http, vertx]
----

[NOTE]
====
Note that the SWUrl was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` when the application started, and will be used at runtime any time it needs it.
====

== Discovery Precedence

Depending on the resource to be discovered, there are some paths to follow, please see the following image for more details:

image::cloud/sw-discovery-flow.jpg[]

> 

Reviewer: Can you include the "Additional Resources" section?

Author: what do you suggest to add as additional resource?

Reviewer: This one at least: https://github.com/kiegroup/kogito-docs/blob/main/serverlessworkflow/modules/ROOT/pages/service-orchestration/configuring-openapi-services-endpoints.adoc

Author: added, thanks.

= Kubernetes Service Discovery for Workflow Application Configuration
:compat-mode!:
// Metadata:
:description: Explain what is and how the service discovery works
:keywords: kogito, workflow, quarkus, serverless, service-discovery, enhanced-service-discovery
// links
:quarkus_issue_url: https://github.com/quarkusio/quarkus/issues/27457

The Kubernetes Service Discovery allows you to have a static URI that will define the Kubernetes resource that you use

> to perform HTTP requests, the following sketch demonstrates how this URI is composed:

Reviewer: In this first paragraph, we should make it crystal clear that the service discovery feature kicks for all URIs that have the supported scheme in the Quarkus configuration properties values.

* **kubernetes**
* **openshift**
* **knative**

For the GVK, the following Kubernetes Resources are currently supported:

* **v1/service**
* **serving.knative.dev/v1/service**
* **v1/pod**
* **apps/v1/deployment**
* **apps.openshift.io/v1/deploymentconfig**
* **apps/v1/statefulset**
* **route.openshift.io/v1/route**
* **networking.k8s.io/v1/ingress**

[IMPORTANT]
====
The **GVK** (Group, Version and Kind) and the **Name** values are **mandatory** and can't be empty, otherwise, the engine will throw a validation issue and the discovery will not be performed.  +
The **namespace** is **optional**. In case of an empty value, the current namespace or context will be used.
====

== Configuration

You can configure the Service Discovery with two **build time** properties, which are:

* `quarkus.kogito.sw.discovery.enabled`: Enable the service discovery. Defaults to _true_.
* `quarkus.kogito.kubernetes-client.okhttp.static.logger`: Weather or not to print the `http` communication between the client and the Kubernetes API. Defaults to _false_.

== How does it work?

During application startup, the service discovery is performed at the Quarkus _STATIC_INIT_ time. First, it scans +

> the Quarkus configuration and searches for the Kubernetes URI pattern. Then if at least one is found, the engine +

Reviewer: ```suggestion
the Quarkus configuration values and searches for the Kubernetes URI pattern. Then if at least one is found, the engine +
```

will parse the URI, query the Kubernetes API for the given resource, and override the given system property.

Suppose we have an application that will consume a resource running on Kubernetes, this resource is a Knative Service +

> that exposes a function, that can be discovered with the following SWUrl:

Reviewer: ```suggestion
that exposes a function that can be discovered with the following URI:
```

2022-08-23 12:21:06,375 INFO  [org.kie.kog.sw.AppLifecycle] (main) The application is starting
2022-08-23 12:21:06,382 INFO  [org.kie.kog.add.qua.mes.com.QuarkusKogitoExtensionInitializer] (main) Registered Kogito CloudEvent extension
2022-08-23 12:21:06,460 INFO  [io.quarkus] (main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Profile prod activated.
2022-08-23 12:21:06,462 INFO  [io.quarkus] (main) Installed features: [cache, cdi, hibernate-validator, jackson-jq, kogito-addon-kubernetes-extension, kogito-addon-messaging-extension, kogito-processes, kogito-serverless-workflow, kubernetes, kubernetes-client, openshift-client, qute, reactive-routes, rest-client, rest-client-jackson, resteasy, resteasy-jackson, smallrye-context-propagation, smallrye-openapi, smallrye-reactive-messaging, smallrye-reactive-messaging-http, vertx]
----

[NOTE]
====

> Note that the SWUrl was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` +

Reviewer: ```suggestion
Note that the URI was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` +
```

when the application started, and will be used at runtime any time it needs it.
====

== Discovery Precedence

Depending on the resource to be discovered, there are some paths to follow, please see the following image for more details:

image::cloud/sw-discovery-flow.jpg[]

== Startup time: discovery enabled x disabled

As the Service Discovery scans the Quarkus Configurations during the startup, it can cause a small delay as shown below:

[tabs]
====
Service Discovery Enabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
----
Service Discovery Disabled::
+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 1.507s. Listening on: http://0.0.0.0:8080
----
====

[IMPORTANT]
====
When using this feature, you need to balance if the delayed startup time is something

> that your application can afford.

Reviewer: Is it possible to implement lazy load? So the first time the application is called, the engine performs the discovery? If so we can have a new JIRA :P

Author: I am not sure, since the QuarkusConfig with interceptor is called at static init and when the system property is called for the first time.
I've opened this jira for further investigation.
https://issues.redhat.com/browse/KOGITO-7873
